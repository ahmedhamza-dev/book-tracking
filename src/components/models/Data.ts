export interface BookData{
    id: number,
    title: string,
    authors: string[],
    shelf?: string,
    imageLinks: {smallThumbnail: string, thumbnail?: string}
}