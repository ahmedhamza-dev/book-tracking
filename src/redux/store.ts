import { configureStore } from "@reduxjs/toolkit";
import bookSlice from "./slices";

const store = configureStore({
    reducer: {
        allBooks: bookSlice.reducer
    }
});

export default store;