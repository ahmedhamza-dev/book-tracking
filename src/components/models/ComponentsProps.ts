import { BookData } from "./Data";

export interface BookProps{
    book: BookData,
};
export interface ShelfProps {
    title: string,
    category: string,
    books: BookData[],
}
export interface SearchShelfProps{
    books: BookData[],
}