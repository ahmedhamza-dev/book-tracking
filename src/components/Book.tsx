import React from "react";
import { useDispatch } from "react-redux";
import { shelfChange } from "../redux/slices";
import { BookProps } from "./models/ComponentsProps";

const Book = ({ book }: BookProps) => {
  const dispatch = useDispatch<any>();
  const updateShelf =  (event: React.ChangeEvent<HTMLSelectElement>) => {
    dispatch(shelfChange({ ...book, shelf: event.target.value }));
  };
  return (
    <li>
      <div className="book">
        <div className="book-top">
          <div
            className="book-cover"
            style={{
              width: 128,
              height: 193,
              backgroundImage: `url(${book.imageLinks.smallThumbnail})`,
            }}
          />
          <div className="book-shelf-changer">
            <select
              onChange={updateShelf}
              value={book.shelf ? book.shelf : "none"}
              data-testid="select"
            >
              <option value="move" disabled>
                Move to...
              </option>
              <option value="currentlyReading">Currently Reading</option>
              <option value="wantToRead">Want to Read</option>
              <option value="read">Read</option>
              <option value="none">None</option>
            </select>
          </div>
        </div>
        <div className="book-title" data-testid="book-title">{book.title}</div>
        <div className="book-authors">
          {book.authors.map((auther: any, index: number) => (
            <span key={index}>
              {auther}
              {index !== book.authors.length - 1 ? <span>, </span> : ""}
            </span>
          ))}
        </div>
        <div className="book-authors">
          <strong data-testid="book-shelf">{book.shelf ? book.shelf : "none"}</strong>
        </div>
      </div>
    </li>
  );
};

export default Book;
