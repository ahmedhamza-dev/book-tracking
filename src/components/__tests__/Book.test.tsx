import React from "react";
import Book from "../Book";
import renderer from "react-test-renderer";
import { render, cleanup } from "@testing-library/react";
import { BookData } from "../models/Data";
import { Provider } from "react-redux";
import store from "../../redux/store";

const book: BookData = {
  id: 1,
  title: "The Linux Command Line",
  authors: ["William E. Shotts, Jr."],
  imageLinks: {
    smallThumbnail:
      "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api",
  },
  shelf: "currentlyReading",
};

afterEach(() => {
  cleanup();
});

it("Book component test", () => {
  const bookComponent = (
    <>
      <Provider store={store}> 
        <Book book={book} />
      </Provider>
    </>
  );
  const component = renderer.create(bookComponent);
  const { getByTestId } = render(bookComponent);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const title = getByTestId("book-title").textContent;
  expect(title).toBe("The Linux Command Line");

  const shelf = getByTestId("book-shelf").textContent;
  expect(shelf).toBe("currentlyReading");
});
