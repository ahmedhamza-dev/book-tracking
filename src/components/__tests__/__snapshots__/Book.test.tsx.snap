// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book component test 1`] = `
<li>
  <div
    className="book"
  >
    <div
      className="book-top"
    >
      <div
        className="book-cover"
        style={
          Object {
            "backgroundImage": "url(http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api)",
            "height": 193,
            "width": 128,
          }
        }
      />
      <div
        className="book-shelf-changer"
      >
        <select
          data-testid="select"
          onChange={[Function]}
          value="currentlyReading"
        >
          <option
            disabled={true}
            value="move"
          >
            Move to...
          </option>
          <option
            value="currentlyReading"
          >
            Currently Reading
          </option>
          <option
            value="wantToRead"
          >
            Want to Read
          </option>
          <option
            value="read"
          >
            Read
          </option>
          <option
            value="none"
          >
            None
          </option>
        </select>
      </div>
    </div>
    <div
      className="book-title"
      data-testid="book-title"
    >
      The Linux Command Line
    </div>
    <div
      className="book-authors"
    >
      <span>
        William E. Shotts, Jr.
        
      </span>
    </div>
    <div
      className="book-authors"
    >
      <strong
        data-testid="book-shelf"
      >
        currentlyReading
      </strong>
    </div>
  </div>
</li>
`;
